instrumentation/dubbo/src/main/java/brave/dubbo/DubboClientRequest.java

```diff
@@ -13,47 +13,62 @@
  */
 package brave.dubbo;
 
-import brave.propagation.Propagation.Setter;
+import brave.Span;
 import brave.rpc.RpcClientRequest;
 import java.util.Map;
+import org.apache.dubbo.common.URL;
 import org.apache.dubbo.rpc.Invocation;
+import org.apache.dubbo.rpc.Invoker;
+import org.apache.dubbo.rpc.RpcContext;
 
-// intentionally not yet public until we add tag parsing functionality
-final class DubboClientRequest extends RpcClientRequest {
-  static final Setter<DubboClientRequest, String> SETTER =
-    new Setter<DubboClientRequest, String>() {
-      @Override public void put(DubboClientRequest request, String key, String value) {
-        request.putAttachment(key, value);
-      }
-
-      @Override public String toString() {
-        return "DubboClientRequest::putAttachment";
-      }
-    };
-
+/** @since 5.12 */
+public final class DubboClientRequest extends RpcClientRequest {
+  final Invoker<?> invoker;
   final Invocation invocation;
   final Map<String, String> attachments;
 
-  DubboClientRequest(Invocation invocation, Map<String, String> attachments) {
+  DubboClientRequest(Invoker<?> invoker, Invocation invocation) {
+    if (invoker == null) throw new NullPointerException("invoker == null");
     if (invocation == null) throw new NullPointerException("invocation == null");
+    this.invoker = invoker;
     this.invocation = invocation;
-    if (attachments == null) throw new NullPointerException("attachments == null");
-    this.attachments = attachments;
+    // When A service invoke B service, then B service then invoke C service, the parentId of the
+    // C service span is A when read from invocation.getAttachments(). This is because
+    // AbstractInvoker adds attachments via RpcContext.getContext(), not the invocation.
+    // See org.apache.dubbo.rpc.protocol.AbstractInvoker(line 141) from v2.7.3
```

Comment:
this is crazy, but it is moved from the old code. ideally there was a bug fixed at some point!